### PR TITLE
Create VM, Storage - force user to choose a storage domain for template disks if necessary

### PR DESCRIPTION
**Pull request by bamsalem**
Fixes: #1230 

Hi I added validation to storage domain for template defined disks.
in case the template's disk should be located in storage domain which not allowed by the current user permissions the disks table in the Create Vm Wizard will be displayed with 'UNKNOWN' storage domain:
image

The user can create this vm with no validation or Error handling.
My new feature will prevent them from create the vm and force them to choose new storage domain according to their permissions, it will disable the next button as long as the user didn't change the storage domain:
Dropdown:
image
Dropdown options:
image
After choosing:
image
What do you think?

### CODE REVIEW

**Comment by ovirt-infra** [comment]
Hello contributor, thanks for submitting a PR for this project! 

I am the bot who triggers "standard-CI" builds for this project.
As a security measure, I will not run automated tests on PRs that are not from white-listed contributors.

In order to allow automated tests to run, please ask one of the project maintainers to review the code and then do one of the following:

1. Type `ci test please` on this PR to trigger automated tests for it.
2. Type `ci add to whitelist` on this PR to trigger automated tests for it and also add you to the contributor white-list so that your future PRs will be tested automatically. ( keep in mind this list might be overwritten if the job XML is refreshed, for permanent whitelisting, please follow #3 option )
3. If you are planning to contribute to more than one project, maybe it's better to ask them to add you to the project organization, so you'll be able to run tests for all the organization's projects.

**Comment by hstastna** [comment]
@bamsalem There are some conflicts to resolve in this PR. Thanks! :)

File: src/components/CreateVmWizard/CreateVmWizard.js
Comment (rszwajko) [inline src/components/CreateVmWizard/CreateVmWizard.js:401]: instead of dedicated flag we can compare indices i.e. `moveBackward = nextStepIndex < activeStepIndex`

File: src/components/CreateVmWizard/CreateVmWizard.js
Comment (rszwajko) [inline src/components/CreateVmWizard/CreateVmWizard.js:181]: `valid` seems always `true` here

File: src/components/CreateVmWizard/steps/Storage.js
Comment (bamsalem) [reply src/components/CreateVmWizard/steps/Storage.js:247]: Hi Radek,
let's take a look on the storage domain `valueView` function:
https://github.com/oVirt/ovirt-web-ui/blob/3a5e6bf5ca0a8c76ac961fcb67435efe68e13512/src/components/CreateVmWizard/steps/Storage.js#L225-L255
I want to focus on the use of the `sdErrFlag` - line 232, the if statement:
if we have case of invalid storage domain the `sd.isOk` will be `undefined` otherwise it will be true even it was invalid first and then the user changed it, if we don't use the `sdErrFlag` in the if statement the user could change it once and then it will be unchangeable. so to avoid this I used a dedicated flag to signify that the template defined disk was invalid for the current user, so this flag should be always true if the disk's SD was invalid at first, otherwise it will change once.
that's the purpose of using always `true` for this flag :)

File: src/components/CreateVmWizard/CreateVmWizard.js
Comment (bamsalem) [reply src/components/CreateVmWizard/CreateVmWizard.js:189]: I don't know why it's not part of the state, maybe we should consider to change it.
@sgratch, @sjd78 what do you think? should we change the `preventExit` to be part of the component's state?

File: src/components/CreateVmWizard/CreateVmWizard.js
Comment (bamsalem) [reply src/components/CreateVmWizard/CreateVmWizard.js:258]: Done. Thanks :)

File: src/components/CreateVmWizard/CreateVmWizard.js
Comment (bamsalem) [reply src/components/CreateVmWizard/CreateVmWizard.js:201]: It's the default value.
maybe I'm not understanding what's the problem?

Comment (rszwajko) [reply src/components/CreateVmWizard/CreateVmWizard.js:201]: Sorry for not being clear enough - this comment is linked with my comment [above](https://github.com/oVirt/ovirt-web-ui/pull/1233#discussion_r449618665)

> It seems we should trigger the validations and set here also the valid flag (otherwise it will be always true).

It seems that we should explicitly set `valid` flag there.

File: src/components/CreateVmWizard/steps/Storage.js
Comment (rszwajko) [reply src/components/CreateVmWizard/steps/Storage.js:247]: Thanks for explanation! I would suggest renaming it to sth like `createdFromInvalidTemplate` and adding some comments in the code to explain this use case (that user should always be able to change it).
How about remaining points?

> It seems we should trigger the validations and set here also the valid flag (otherwise it will be always true).
> It seems we don't need to pass update prop name as a string.

Comment (hstastna) [reply src/components/CreateVmWizard/steps/Storage.js:247]: > ...the user could change it once and then it will be unchangeable...

I wonder why it was implemented like that, originally. I think we should solve this the other way than adding an extra flag which will always be set to the same value. But I feel I still don't understand all the details so maybe I just don't see the reason why you've chosen to solve the problem of changing SD by adding such a flag :)

Comment (bamsalem) [reply src/components/CreateVmWizard/steps/Storage.js:247]: Hi Radek,
the validations are in the CreateVmWizard component and the update prop is not string anymore.
Thanks.

File: src/components/CreateVmWizard/CreateVmWizard.js
Comment (bamsalem) [reply src/components/CreateVmWizard/CreateVmWizard.js]: thanks, I don't know why we still can see that , but it's something I already changed

File: src/components/CreateVmWizard/steps/Storage.js
Comment (bamsalem) [reply src/components/CreateVmWizard/steps/Storage.js:233]: I'm not sure what you want me to do here, can you explain me please, thanks :)

**Comment by hstastna** [comment]
@bamsalem There are some conflicts in the _CreateVmWizard.js_ that need to be resolved. Thanks! :)

File: src/components/SelectBox.js
Comment (hstastna) [inline src/components/SelectBox.js:53]: I think we don't need this case right now. I would avoid adding unnecessary code which can be added later anytime if needed. Otherwise, LGTM.

**Review by hstastna — APPROVED** [review]
Except some unnecessary code from my point of view, LGTM.

File: src/components/SelectBox.js
Comment (bamsalem) [reply src/components/SelectBox.js:53]: Hi Hilda, I agree that is not necessary right now, but IMHO it's just component's functionality, so if we already edited the component, it will be better to add the functionality to create consistency with project's UI library - PF3 WDYT?

File: src/components/CreateVmWizard/steps/Storage.js
Comment (rszwajko) [reply src/components/CreateVmWizard/steps/Storage.js:284]: My understanding is that @sjd78  suggests calculating `createdFromInvalidTemplate` upfront in the same way `isFromTemplate` is calculated.

**Review by rszwajko — CHANGES_REQUESTED** [review]

File: src/components/CreateVmWizard/steps/Storage.js
Comment (bamsalem) [reply src/components/CreateVmWizard/steps/Storage.js:295]: thanks @rszwajko ,I did it.

File: src/components/SelectBox.js
Comment (hstastna) [reply src/components/SelectBox.js:53]: I still think we should avoid adding unnecessary code. We don't know if we will use it in the future.

**Review by rszwajko — APPROVED** [review]

File: src/components/CreateVmWizard/steps/Storage.js
Comment (sjd78) [reply src/components/CreateVmWizard/steps/Storage.js:311]: Or just remove the comment since I think the condition on line 229 is sufficiently descriptive.

File: src/components/SelectBox.js
Comment (sjd78) [inline src/components/SelectBox.js:97]: Looks like `validationState` should just be a list of possible values, and have it be optional (default 'default').  So like:

```jsx
validationState: PropTypes.oneOf([ 'default', 'error' ]),
```

Then you can add a default value like this (add below the propTypes):

```jsx
SelectBox.defaultProps = {
  validationState: 'default',
}
```

File: src/components/SelectBox.js
Comment (sjd78) [inline src/components/SelectBox.js:47]: Add a blank line between component methods.

I'd just call it `getValidationClass`.

File: src/components/CreateVmWizard/steps/Storage.js
Comment (sjd78) [reply src/components/CreateVmWizard/steps/Storage.js:311]: Or just remove the comment since I think the condition on line 229 is sufficiently descriptive.

**Review by sjd78 — CHANGES_REQUESTED** [review]
A few minor clean up comments and it'll be all set.

File: src/components/SelectBox.js
Comment (bamsalem) [reply src/components/SelectBox.js:97]: @sjd78 it seems like it doesn't work in that case:
image
In case of using `OR`/`AND` statements `false` is possible returned value, so we need to consider it in our `PropTypes`.
WDYT about using one of the options below:
```
validationState: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.oneOf([ false ]),
  ]),
```
```
  validationState: PropTypes.oneOf([ false, 'default', 'error' ]),
```
Do you have another suggestion?

Comment (rszwajko) [reply src/components/SelectBox.js:98]: I understand you refer to following usage pattern:
`validationState={!sd.isOk && 'error'}`
It's an universal problem of dynamic optional attribute. There are several ways to solve that (i.e. [this thread])(https://stackoverflow.com/questions/31163693/how-do-i-conditionally-add-attributes-to-react-components).
Actually I would rather avoid embedding that in type definition (sorry for not spotting that during first review). 
We could i.e. just fall-back to default: `validationState={!sd.isOk && 'error' || 'default'}`

Comment (bamsalem) [reply src/components/SelectBox.js:98]: I prefer the trinary statement too, in my first PRs @sjd78 ask me every time to change trinary statements to that pattern and in the project most of the time we are using that pattern:
> validationState={!sd.isOk && 'error'}

So IMHO we need to support using this pattern or change it everywhere in the project to keep consistency.
WDYT?

Comment (hstastna) [reply src/components/SelectBox.js:98]: I think both options are okay if t's just about PropTypes. It doesn't have any impact on functionality, right?

Comment (sjd78) [reply src/components/SelectBox.js:98]: Ok, so....

While it is true that PropTypes don't break things at run time, we should not be seeing any errors when running in dev mode.  In prod mode, PropTypes warnings are not added.  Runtime type errors tend to indicate that something will go very wrong, very quietly.

Yes I usually like to see a `condition && something` rather than `condition ? something : undefined` but it is more context sensitive.  When doing components that are optionally rendered, the `condition && <Component>` is my favorite.  When doing attributes, could go either way, but really depends on what the attribute values need to be.  With an attribute that needs to be `undefined` (PropTypes not required) or one of a set of possible values (PropTypes.oneOf), the ternary makes sense.  So here, it would look like `validationState={sd.isOk ? 'default' : 'error'}` if the PropTypes doesn't allow a false value that maps to 'default'.

If `SelectBox` will only have 2 states - normal and error - then the attribute can just be a boolean `valid`.  If we want to have >2 states, then a enum of strings makes sense.  That just means the optional attribute must be present and a string or must be undefined and then get a default value.

Finally, using: `PropTypes.oneOf([ false, 'default', 'error' ])` and handling `false` the same as `undefined` / use the default value, should be ok.

**Review by sjd78 — APPROVED** [review]
LGTM

**Comment by sgratch** [comment]
@bamsalem When verifying this PR, I'm not forced to choose a valid storage domain for a user when creating a VM.
For both template's and new disks there is no verification that the disk’s storage domain is a valid domain for the current user. It's just failed to create on backend due to permissions issue ("Cannot add VM. The user doesn't have permissions to attach Disk Profile to the Disk").
Am I missing something?

**Comment by sgratch** [comment]
When verifying this PR, I'm not forced to choose a valid storage domain for a user when creating a VM.
> For both template's and new disks there is no verification that the disk’s storage domain is a valid domain for the current user. It's just failed to create on backend due to permissions issue ("Cannot add VM. The user doesn't have permissions to attach Disk Profile to the Disk").

This is actually divided into 2 sub issues:

- **For creating a disk based on a template** - this is blocked by #1237 since currently the option to create a disk based on a template is not fully supported.  Once #1237 is done, we'll need to enable changing the SD and disk type for template's disks for supporting this issue. Or we can display an error and avoid the user from creating such a VM. 

- **For creating a disk not based on a template**, when creating a disk for a user that has no "_diskprofileuser_" permissions for the SD, the ui allows me to choose that SD and the disk creation failed only on backend due to permissions issue.

  It seems that the logged user should have permissions for the SD and also should be allowed to manage disks on that SD.
Do we verify both with the API query?

**Comment by sjd78** [comment]
> When verifying this PR, I'm not forced to choose a valid storage domain for a user when creating a VM.
> 
> > For both template's and new disks there is no verification that the disk’s storage domain is a valid domain for the current user. It's just failed to create on backend due to permissions issue ("Cannot add VM. The user doesn't have permissions to attach Disk Profile to the Disk").
> 
> @bamsalem @sjd78 This is actually divided into 2 sub issues, both are not working:
> 
>   * **For creating a disk based on a template** - this is blocked by #1237 since currently the option to create a disk based on a template is not fully supported.  Once #1237 is done, we'll need to enable changing the SD and disk type for template's disks for supporting this issue. Or we can display an error and avoid the user from creating such a VM.

This PR adds the SD selection, but the create VM saga was not adjusted to handle the change.  My PR #1316, which fixes the types and updates the VM create saga, hopefully handles the saga changes needed in this PR.  I will need to do some extra testing with both PRs applied to verify.

> 
>   * **For creating a disk not based on a template**, when creating a disk for a user that has no "_diskprofileuser_" permissions for the SD, the ui allows me to choose that SD and the disk creation failed only on backend due to permissions issue.
>       It seems that the logged user should have permissions for the SD and also should be allowed to manage disks on that SD.
>       Do we verify both with the API query?

The list of storage domains shown to the user is based on `canUserUseDomain` calculated in the saga that fetches the storage domain: https://github.com/oVirt/ovirt-web-ui/blob/124d2858c7f8a7867d27d9a6fcd5bcd8d9a03660/src/sagas/storageDomains.js#L67-L85

Remember, we check based on permit and not on permission, and `canUserUseDomain` is checking if the user has the permit `create_disk`.  It is quite possible a user picks up the necessary permit via another permission/role.  See: https://github.com/oVirt/ovirt-web-ui/blob/124d2858c7f8a7867d27d9a6fcd5bcd8d9a03660/src/utils/permissions.js#L33-L35

Sounds like we'll need to adjust the permit checking and add a `canUserCreateDisks` type of check and use that in various places.  Or at least update `canUserUseDomain`'s check.

**Comment by sjd78** [comment]
This PR is to be replaced with #1323.  It is this PR merged with #1316.

**Comment by sjd78** [comment]
> > * **For creating a disk based on a template** - this is blocked by #1237 since currently the option to create a disk based on a template is not fully supported.  Once #1237 is done, we'll need to enable changing the SD and disk type for template's disks for supporting this issue. Or we can display an error and avoid the user from creating such a VM.
> 
> This PR adds the SD selection, but the create VM saga was not adjusted to handle the change. My PR #1316, which fixes the types and updates the VM create saga, hopefully handles the saga changes needed in this PR. I will need to do some extra testing with both PRs applied to verify.

@sgratch , @bamsalem - This is now working in every case that I tested.  There may be some other edge cases, but at this point I'm not sure it is worth finding and handling them.  We should be well covered as is over in #1323.

> 
> > * **For creating a disk not based on a template**, when creating a disk for a user that has no "_diskprofileuser_" permissions for the SD, the ui allows me to choose that SD and the disk creation failed only on backend due to permissions issue.
> >   It seems that the logged user should have permissions for the SD and also should be allowed to manage disks on that SD.
> >   Do we verify both with the API query?
> 
> Remember, we check based on permit and not on permission, and `canUserUseDomain` is checking if the user has the permit `create_disk`. It is quite possible a user picks up the necessary permit via another permission/role. See:
> 
> Sounds like we'll need to adjust the permit checking and add a `canUserCreateDisks` type of check and use that in various places. Or at least update `canUserUseDomain`'s check.

@sgratch , @bamsalem - The storage domain drop down was displaying the storage domains that I expected to see when unit testing #1323.  It was not a comprehensive test, more of an average path set of tests.  I would not hold up #1323 (which will take the place of this PR) with additional permit checking concerns.

**Comment by sjd78** [comment]
Moved the last comment to #1323, closing this PR.

**Pull request closed without merging**
